File: PAT_Basic_Level_by_java/1044 火星数字/1044 火星数字.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

/**
 * 1044 火星数字
 *
 * 题目描述：
 * 实现十进制数（范围 [0, 169)）与火星十三进制表示之间的相互转换。火星文使用
 * 两个词表：低位词（"tret", "jan", "feb", ... 对应 0~12）和高位词（"tam",
 * "hel", ... 对应 13~12×13）。
 *
 * 实现原理：
 * 1. 十进制 → 火星文：
 *    - 0 特殊处理返回 "tret"
 *    - 按 13 进制分解为高位 (num/13) 和低位 (num%13)
 *    - 高位为 0 只输出低位词，低位为 0 只输出高位词，否则输出 "高位词 低位词"
 * 2. 火星文 → 十进制：
 *    - 按空格分割，对每个 token 分别在 HIGH 和 LOW 词表中查找
 *    - 在高位词表中找到则累加 13*h，在低位词表中找到则累加 l
 * 3. indexOf 方法实现线性查找词表。
 *
 * 时间复杂度：O(N)，N 为查询次数。每次转换 O(1)
 */
public class Main {
    private static final String[] LOW = {"tret", "jan", "feb", "mar", "apr", "may", "jun",
            "jly", "aug", "sep", "oct", "nov", "dec"};
    private static final String[] HIGH = {"", "tam", "hel", "maa", "huh", "tou", "kes",
            "hei", "elo", "syy", "lok", "mer", "jou"};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine().trim());
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < n; i++) {
            String line = br.readLine().trim();
            if (Character.isDigit(line.charAt(0))) {
                sb.append(toMars(Integer.parseInt(line))).append('\n');
            } else {
                sb.append(toNum(line)).append('\n');
            }
        }
        System.out.print(sb);
    }

    // 十进制 → 火星文：按 13 进制分解为高位和低位，查表转换
    private static String toMars(int num) {
        if (num == 0) return "tret";
        int high = num / 13;   // 高位（13的倍数）
        int low = num % 13;    // 低位（0~12）
        if (high == 0) return LOW[low];         // 只有低位
        if (low == 0) return HIGH[high];        // 只有高位（整十位）
        return HIGH[high] + " " + LOW[low];     // 高位 + 低位
    }

    // 火星文 → 十进制：按空格分割后分别在 HIGH 和 LOW 词表中查找
    private static int toNum(String s) {
        String[] parts = s.split("\\s+");
        int val = 0;
        for (String p : parts) {
            int h = indexOf(HIGH, p);
            if (h > 0) val += h * 13;   // 在高位词表中找到
            else val += indexOf(LOW, p); // 在低位词表中找到
        }
        return val;
    }

    // 在字符串数组中线性查找指定值，返回下标
    private static int indexOf(String[] arr, String s) {
        for (int i = 0; i < arr.length; i++) {
            if (arr[i].equals(s)) return i;
        }
        return -1;
    }
}

File: PAT_Basic_Level_by_py3/1044 火星数字/1044 火星数字.py
# 题目：1044 火星数字
#
# 题目描述：
#   火星人是以 13 进制计数的：
#   - 地球人的 0 被火星人称为 tret。
#   - 地球人数字 1 到 12 的火星文分别为：jan, feb, mar, apr, may, jun, jly, aug, sep, oct, nov, dec。
#   - 火星人将进位以后的 12 个高位数字分别称为：tam, hel, maa, huh, tou, kes, hei, elo, syy, lok, mer, jou。
#   例如地球人的数字 `29` 翻译成火星文就是 `hel mar`；而火星文 `elo nov` 对应地球数字 `115`。为了方便交流，请你编写程序实现地球和火星数字之间的互译。
#
# 输入格式：
#   输入第一行给出一个正整数 N（<100），随后 N 行，每行给出一个 [0, 169) 区间内的数字 —— 或者是地球文，或者是火星文。
#
# 输出格式：
#   对应输入的每一行，在一行中输出翻译后的另一种语言的数字。
#
# 实现原理：
#   - 循环迭代处理
#   - 列表操作
#
# 算法思路：
#   进制与字符串转换：实现地球数字与火星数字的相互转换，
#   处理13进制和特殊的火星文表示。
#
import sys

low = 'tret jan feb mar apr may jun jly aug sep oct nov dec'.split(); high = 'tam hel maa huh tou kes hei elo syy lok mer jou'.split()
for _ in range(int(sys.stdin.readline())):
    s = sys.stdin.readline().strip()
    if s[0].isdigit():
        n = int(s); print(high[n//13] + ((' ' + low[n%13]) if n%13 else '') if n >= 13 else low[n])
    else:
        a = s.split(); print((high.index(a[0])*13 if a[0] in high else 0) + (low.index(a[-1]) if a[-1] in low else 0))
